### ISSUE TITLE
canbus support (QEMU-209)

### ISSUE THREAD

**Issue by maedoc** [espressif/qemu#101]
### Checklist

- [X] Checked the issue tracker for similar issues to ensure this is not a duplicate.
- [X] Described the feature in detail and justified the reason for the request.
- [X] Provided specific use cases and examples.

### Feature description

Support the canbus object per https://www.qemu.org/docs/master/system/devices/can.html

### Use cases

I would like to debug my esp32 app using the twai (or similar) with a virtual can interface on linux

### Alternatives

The linux target, but it doesn't compile with my app.

**Comment by igrr** [comment]
@maedoc I've looked into the possibility of CAN support, and it seems very doable. QEMU already has an implementation of SJA1000 controller, we just need to implement a small wrapper device to attach SJA1000 as an MMIO peripheral (similar to hw/net/can/can_pcm3680_pci.c, just without PCI...)

This will only work on Linux hosts, though, since the only CAN host in QEMU is `socketcan`, which is only supported on Linux. I would just like to confirm whether you are okay with using Linux as host OS?

**Comment by maedoc** [comment]
> okay with using Linux as host OS? 

yes only been using esp idf on Linux, or Docker on macOS, so that would be fine.

thanks for considering implementing!

**Comment by wodz** [comment]
Here you can find patches which implement CAN (TWAI) esp32 peripheral. (I had to change extension to .txt because of some stupid github attachements policy).

1) In order to share canbus with host you need to have real socket can interface or virtual one up. To setup virtual can interface run on the host:
modprobe vcan
ip link add dev vcan0 type vcan
ip link set vcan0 mtu 16
ip link set up vcan0

2) Add params when running patched qemu-system-xtensa:
-object can-bus,id=canbus0 -object can-host-socketcan,id=canhost0,if=vcan0,canbus=canbus0 -global driver=esp32.twai,property=canbus,value=canbus0

Enjoy.

[0001-Implement-esp32-TWAI.txt](https://github.com/user-attachments/files/18506213/0001-Implement-esp32-TWAI.txt)

[0003-cherry-pick-25145a7d77-from-upstream.txt](https://github.com/user-attachments/files/18506221/0003-cherry-pick-25145a7d77-from-upstream.txt)
[0004-SJA1000-fix-RMC-read.txt](https://github.com/user-attachments/files/18506222/0004-SJA1000-fix-RMC-read.txt)

**Comment by maedoc** [comment]
hi @wodz thanks for the patches.  I will try them shortly.

**Comment by eternal-echo** [comment]
Hi @maedoc (and thanks to @wodz for the initial patches!),

I’ve now adapted and ported the [ESP32 TWAI CAN wrapper into QEMU,](https://github.com/eternal-echo/qemu/tree/esp-qemu) based on @wodz’s work, and verified it works end-to-end on Linux. Here’s how to try it out:

[flash_image.zip](https://github.com/user-attachments/files/21037354/flash_image.zip)

[qemu-system-xtensa.zip](https://github.com/user-attachments/files/21038011/qemu-system-xtensa.zip)


1. **Prepare a virtual CAN interface**  
    ```bash
    sudo modprobe vcan
    sudo ip link add dev vcan0 type vcan
    sudo ip link set vcan0 up
    ```

2. **Configure & build QEMU**  
    ```bash
    mkdir build && cd build
    ../configure \
      --target-list=xtensa-softmmu \
      --enable-gcrypt \
      --enable-slirp \
      --enable-debug \
      --enable-sdl \
      --disable-strip \
      --disable-user \
      --disable-capstone \
      --disable-vnc \
      --disable-gtk \
      --disable-docs \
      --enable-trace-backends=log
    ninja
    ```

3. **Run QEMU with TWAI attached to vcan0**  
    you can unzip my test demo for twai.
    ```bash
    build/qemu-system-xtensa -nographic \
      -machine esp32 \
      -drive file=flash_image.bin,if=mtd,format=raw \
      -object can-bus,id=canbus0 \
      -object can-host-socketcan,id=canhost0,if=vcan0,canbus=canbus0 \
      -global driver=esp32.twai,property=canbus,value=canbus0 \
      -d guest_errors,unimp \
      -D qemu.log
    ```

4. **Send/receive CAN frames from the host**  
    ```bash
    # Listen for all frames
    candump vcan0

    # Send a standard (11-bit) frame ID=0x123, data=0x11 22 33 44 55 66
    cansend vcan0 123#112233445566
    ```

With this setup, you should see CAN frames flowing between your ESP32 firmware in QEMU and the host’s vcan0 interface. All core TWAI registers, interrupts, and SJA1000 features are exercised and tested.

Let me know how it goes or if you hit any issues!

**Comment by maedoc** [comment]
*editing my response yet again after some sleep & more patience*

Thanks again for looking at this request.  I was able to build qemu from the branch linked above, and successfully use this with a single can device under the v2 API. at least once in any case..
